src/Compilers/Core/Portable/Compilation/Expression.cs

```diff
@@ -0,0 +1,380 @@
+            }
+
+            if (type.TypeKind == TypeKind.Enum)
+            {
+                return SynthesizeNumeric(((INamedTypeSymbol)type).EnumUnderlyingType, value);
+            }
+
+            return null;
+        }
+
+        public static BinaryOperationKind DeriveAdditionKind(ITypeSymbol type)
+        {
+            switch (type.SpecialType)
+            {
+                case SpecialType.System_Int32:
+                case SpecialType.System_Int64:
+                case SpecialType.System_Int16:
+                case SpecialType.System_SByte:
+                    return BinaryOperationKind.IntegerAdd;
+                case SpecialType.System_UInt32:
+                case SpecialType.System_UInt64:
+                case SpecialType.System_UInt16:
+                case SpecialType.System_Byte:
+                case SpecialType.System_Char:
+                case SpecialType.System_Boolean:
+                    return BinaryOperationKind.UnsignedAdd;
+                case SpecialType.System_Single:
+                case SpecialType.System_Double:
+                    return BinaryOperationKind.FloatingAdd;
+                case SpecialType.System_Object:
+                    return BinaryOperationKind.ObjectAdd;
+            }
+
+            if (type.TypeKind == TypeKind.Enum)
+            {
+                return Semantics.BinaryOperationKind.EnumAdd;
+            }
+
+            return Semantics.BinaryOperationKind.None;
+        }
+
+        public static LiteralKind DeriveLiteralKind(ITypeSymbol type)
+        {
+            switch (type.SpecialType)
+            {
+                case SpecialType.System_Boolean:
+                    return Semantics.LiteralKind.Boolean;
+                case SpecialType.System_Byte:
+                case SpecialType.System_Int16:
+                case SpecialType.System_Int32:
+                case SpecialType.System_Int64:
+                case SpecialType.System_IntPtr:
+                case SpecialType.System_SByte:
+                case SpecialType.System_UInt16:
+                case SpecialType.System_UInt32:
+                case SpecialType.System_UInt64:
+                case SpecialType.System_UIntPtr:
+                    return Semantics.LiteralKind.Integer;
+                case SpecialType.System_DateTime:
+                    return Semantics.LiteralKind.DateTime;
+                case SpecialType.System_Decimal:
+                    return Semantics.LiteralKind.Decimal;
+                case SpecialType.System_Double:
+                case SpecialType.System_Single:
+                    return Semantics.LiteralKind.Floating;
+                case SpecialType.System_Char:
+                    return Semantics.LiteralKind.Character;
+                case SpecialType.System_String:
+                    return Semantics.LiteralKind.String;
+            }
+
+            return Semantics.LiteralKind.None;
+        }
+    }
+
+    public class Relational : IRelational
```

Comment:
Are we using these non-BoundNode implementations of IExpression? Ideally we should simply use BoundNodes rather than creating parallel trees or sub-trees.